tree/ntuple/v7/test/ntuple_bulk.cxx

```diff
@@ -86,3 +86,99 @@ TEST(RNTupleBulk, Complex)
+
+   auto fieldZero = reader->GetModel()->GetFieldZero();
+   std::unique_ptr<RFieldBase::RBulk> bulk32;
+   std::unique_ptr<RFieldBase::RBulk> bulk64;
+   for (auto &f : *fieldZero) {
+      if (f.GetName() == "card32")
+         bulk32 = std::make_unique<RFieldBase::RBulk>(f.GenerateBulk());
+      if (f.GetName() == "card64")
+         bulk64 = std::make_unique<RFieldBase::RBulk>(f.GenerateBulk());
+   }
+
+   auto mask = std::make_unique<bool[]>(10);
+   std::fill(mask.get(), mask.get() + 10, false /* the cardinality field optimization should ignore the mask */);
+
+   auto card32Arr = static_cast<std::uint32_t *>(bulk32->ReadBulk(RClusterIndex(0, 0), mask.get(), 10));
+   auto card64Arr = static_cast<std::uint64_t *>(bulk64->ReadBulk(RClusterIndex(0, 0), mask.get(), 10));
+   for (int i = 0; i < 10; ++i) {
+      EXPECT_EQ(i, card32Arr[i]);
+      EXPECT_EQ(i, card64Arr[i]);
+   }
+}
+
+TEST(RNTupleBulk, RVec)
+{
+   FileRaii fileGuard("test_ntuple_bulk_rvec.root");
+   {
+      auto model = RNTupleModel::Create();
+      auto fldVecI = model->MakeField<ROOT::RVecI>("vint");
+      auto fldVecS = model->MakeField<ROOT::RVec<CustomStruct>>("vs");
+      auto writer = RNTupleWriter::Recreate(std::move(model), "ntpl", fileGuard.GetPath());
+      for (int i = 0; i < 10; ++i) {
+         fldVecI->resize(i);
+         fldVecS->resize(i);
+         for (int j = 0; j < i; ++j) {
+            fldVecI->at(j) = j;
+            fldVecS->at(j).a = j;
+         }
+         writer->Fill();
+      }
+   }
+
+   auto reader = RNTupleReader::Open("ntpl", fileGuard.GetPath());
+
+   auto fieldZero = reader->GetModel()->GetFieldZero();
+   std::unique_ptr<RFieldBase::RBulk> bulkI;
+   std::unique_ptr<RFieldBase::RBulk> bulkS;
+   for (auto &f : *fieldZero) {
+      if (f.GetName() == "vint")
+         bulkI = std::make_unique<RFieldBase::RBulk>(f.GenerateBulk());
+      if (f.GetName() == "vs")
+         bulkS = std::make_unique<RFieldBase::RBulk>(f.GenerateBulk());
+   }
+
+   auto mask = std::make_unique<bool[]>(10);
+   std::fill(mask.get(), mask.get() + 10, true);
+   mask[1] = false; // the RVec<simple type> field optimization should ignore the mask
+
+   auto iArr = static_cast<ROOT::RVecI *>(bulkI->ReadBulk(RClusterIndex(0, 0), mask.get(), 10));
+   auto sArr = static_cast<ROOT::RVec<CustomStruct> *>(bulkS->ReadBulk(RClusterIndex(0, 0), mask.get(), 10));
+   for (int i = 0; i < 10; ++i) {
+      EXPECT_EQ(i, iArr[i].size());
+      EXPECT_EQ(i == 1 ? 0 : i, sArr[i].size());
+      for (std::size_t j = 0; j < iArr[i].size(); ++j) {
+         EXPECT_EQ(j, iArr[i].at(j));
+      }
+      for (std::size_t j = 0; j < sArr[i].size(); ++j) {
+         EXPECT_FLOAT_EQ(j, sArr[i].at(j).a);
+      }
+      // RVec<PoD> should have all the vector elements of the bulk stored consecutively in memory
+      if (i > 1) {
+         EXPECT_EQ(&iArr[i - 1][0] + iArr[i - 1].size(), &iArr[i][0]);
+      }
+   }
+}
```

Comment:
how confident are we that given we are testing `RVec<SimpleType>` and `RVec<ComplexType>` we don't need to test `RVec<RVec<SimpleType>>`? :grimacing: